File: hw4/matlab/hw4_iskelet.m
img=imread('HW4_2.tif'); %we are reading image in here
subplot(2,4,1)
imshow(img)
[a,b,c]=size(img);
f=3;
filter=[-1 -1 -1;-1 8 -1;-1 -1 -1];
%filter=(filter)/sum(filter(:));
%filter=double(filter);
array=padarray(img,[1 1]);
array=double(array);
output1=zeros(a,b);
output2=zeros(a,b);
output3=zeros(a,b);
for i=1:size(array,1)-2
    for j=1:size(array,2)-2
    temp1=array(i:i+f-1,j:j+f-1).*filter;
    output1(i,j)=sum(temp1(:));
    end
end
subplot(2,4,2)
imshow(uint8(output1))
image_c=img+uint8(output1);

subplot(2,4,3)
imshow(image_c)

for i=1:size(array,1)-2
    for j=1:size(array,2)-2
       Gx=((2*array(i+2,j+1)+array(i+2,j)+array(i+2,j+2))-(2*array(i,j+1)+array(i,j)+array(i,j+2)));
       Gy=((2*array(i+1,j+2)+array(i,j+2)+array(i+2,j+2))-(2*array(i+1,j)+array(i,j)+array(i+2,j)));
       output2(i,j)=sqrt(Gx.^2+Gy.^2);
    end
end
image_d=uint8(output2);
subplot(2,4,4)
imshow(image_d)
image_e=mean_filter(image_d,5);
image_e=uint8(image_e);
subplot(2,4,5)
imshow(image_e)
image_f=medfilt2(double(image_e) .* double(image_c))/255;
subplot(2,4,6)
imshow(uint8(image_f))
image_g=uint8(output1)+img;
subplot(2,4,7)
imshow(image_g)
gamma=exp(0.1);
r=(double(image_g).^gamma);
subplot(2,4,8)
imshow(uint8(r))

File: hw4/python/hw4_iskelet.py
import math

import cv2
import numpy as np

img=cv2.imread("imagess/HW4_2.tif")
cv2.imshow("original window",img)
img=cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)
[a,b]=img.shape
filter=[[-0.125, -0.125, -0.125],[-0.125 ,1, -0.125],[-0.125, -0.125, -0.125]]
filter=np.double(filter)
array=np.zeros(shape=(a+2,b+2))
array2=np.zeros(shape=(a+2,b+2))
array3=np.zeros(shape=(a+2,b+2))
for i in range(a):
    for j in range(b):
        x=img[i,j]
        array[i+1,j+1]=x
        array2[i + 1, j + 1] = x
        array3[i + 1, j + 1] = x
array=np.double(array)
array2=np.double(array2)
output1=np.zeros([a,b])
output2=np.zeros([a,b])
for i in range(a-2):
     for j in range(b-2):
         temp1=np.multiply(array[i:i+3,j:j+3],filter)
         output1[i,j]=temp1.sum()

image_b=np.uint8(output1)
cv2.imshow("image b window",image_b)

image_c=image_b+img
cv2.imshow("image c window",image_c)

for i in range(1,a-2):
    for j in range(1,b-2):
       Gx=((2*array2[i+2,j+1]+array2[i+2,j]+array2[i+2,j+2])-(2*array2[i,j+1]+array2[i,j]+array2[i,j+2]))
       Gy=((2*array2[i+1,j+2]+array2[i,j+2]+array2[i+2,j+2])-(2*array2[i+1,j]+array2[i,j]+array2[i+2,j]));
       output2[i,j]=math.sqrt(pow(Gx,2)+pow(Gy,2))

image_d=np.uint8(output2)
cv2.imshow("image d window",image_d)
#mean filtering
filter=np.ones([5,5])
filter=(filter)/25
filter=np.double(filter)
array3=np.double(array3)
output3=np.zeros([a,b])
for i in range(a-4):
     for j in range(b-4):
         temp1=np.multiply(array3[i:i+5,j:j+5],filter)
         output3[i,j]=temp1.sum()



image_e=np.uint8(output3)
cv2.imshow("image e window",image_e)
image_f=np.multiply(np.double(image_e),np.double(image_c))
image_f=image_f/255
image_f=np.uint8(image_f)
cv2.imshow("image f window",image_f)
image_g=image_f+img
cv2.imshow("image g window",image_g)
image_h=np.double(image_g)**1.0
cv2.imshow("image h window",np.uint8(image_h))
cv2.waitKey(0)
cv2.destroyAllWindows()